### PR TITLE
Common prefix pipeline computation in pt.Experiment

### PR DESCRIPTION
**Pull request by cmacdonald**
This PR is for precomputation of common interim results by `pt.Experiment`. Consider you are evaluating using `pt.Experiment` on multiple pipelines with exactly the same initial stages. This kwarg option to `pt.Experiment` will precompute the results of the common initial stages, and then use these results to call the subsequent remainder of each pipelines.

For example:
```python
bm25 = pt.terrier.Retriever.from_dataset('vaswani', 'terrier_stemmed', wmodel='BM25')
pipeB = bm25 %10
pt.Experiment(
  [bm25, pipeB], 
  pt.get_dataset('vaswani').get_topics(), 
  pt.get_dataset('vaswani').get_qrels(), 
  eval_metrics=['map'], 
  precompute_shared=True
)
```
will now only invoke `bm25` once on the topics, and then reuse its results for the remainder of `pipeB`, resulting in increased experimentation efficiency.

Integration of this PR will address one of the downsides of using PyTerrier pipelines rather than iterative development.

Issues for discussion, further testing:
- reliability of `__equals__` for `pt.Transformer` objects, particularly `pt.apply` functions?
- naming of the kwarg `precompute_shared` in pt.Experiment
- need for further example documentation in experiments.rst
- impact on "mrt" timings (e.g. we could just emit a warningf)
- impact on batch calculations
- this only works if all pipelines have the same common prefix. Handling any more complex situations (e.g. common prefix of 2/3 pipelines) would, I think, make the code too complex. 
- how could this scale to e.g. thousands of queries?

I need to really test this in anger on a bigger dataset.

Observation: I hadn't attempted this in the past as `a >> b >> c` was always instantiated as `(a >> b) >> c` etc. The v0.12 compilation stuff fixed this - thanks @seanmacavaney!

### CODE REVIEW

**Review by seanmacavaney — COMMENTED** [review]
Awesome! Just one main concern, and a few nits. I vote we make the default True.

> reliability of __equals__ for pt.Transformer objects, particularly pt.apply functions?

IMO: This is just an optimization. If it doesn't work, no harm done. (Unless somebody implements a faulty __equals__, but I think that's beyond the scope of what we can be expected to reasonably handle.)

> naming of the kwarg precompute_shared in pt.Experiment

Not sure on this one. I think a better name probably exists, but I don't know what it is.

> impact on "mrt" timings (e.g. we could just emit a warning)

Time the precomputation and add to the final MRT?

> this only works if all pipelines have the same common prefix. Handling any more complex situations (e.g. common prefix of 2/3 pipelines) would, I think, make the code too complex.

Agreed. This is a great starting point and we can see how far it gets us.

> how could this scale to e.g. thousands of queries?

The concern is that the intermediate representation might be too big to store in memory?

File: pyterrier/pipelines.py [hunk 1]
```diff
@@ -528,7 +611,7 @@ def _apply_round(measure, value):
                 save_file = os.path.join(save_dir, "%s.%s" % (name, save_ext))
 
             time, evalMeasuresDict = _run_and_evaluate(
-                system, topics, qrels, eval_metrics, 
+                system, execution_topics, qrels, eval_metrics,
```
[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier/pipelines.py:614]: I worry about the interaction with large interim result sets

File: pyterrier/pipelines.py [hunk 2]
```diff
@@ -144,6 +145,66 @@ def _ir_measures_to_dict(
...
 
+def _lcsMany(pipes : List[List[pt.Transformer]]) -> Tuple[List[Transformer], List[List[Transformer]]]:
+    assert len(pipes) > 1
+    
+    lens = [len(p) for p in pipes]
+    common_prefix=[]
+    
+    for i in range(min(lens)):
+        pipeFirst = pipes[0][i]
+        # pipeFirst can be a dataframe
```
[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier/pipelines.py:156]: I think that would make changes in other parts of that file, eg _run_and_evaluate

**Comment by cmacdonald** [comment]
> naming of the kwarg precompute_shared

Common or shared?

> The concern is that the intermediate representation might be too big to store in memory?

Yes. Think thousands of topics * hundreds or a thousand results?

File: pyterrier/pipelines.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (seanmacavaney) [reply pyterrier/pipelines.py:614]: Okay, but then I think we’ve gotta do a bit of a crusade on transformers that make changes in-place to their inputs. (It should probably be done anyway.) I’ll write up an issue.

File: pyterrier/pipelines.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (seanmacavaney) [reply pyterrier/pipelines.py:156]: Yeah, that could cause some unintended side-effects. Maybe transformerizing everything at the start of this function instead?

File: tests/test_experiment_precomp.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import os
+import unittest
+import warnings
+import math
+from pyterrier.measures import *
+from .base import BaseTestCase
+import pytest
+
+class TestExperiment(BaseTestCase):
+    def test_common(self):
```
[the remark this answers is not in the document]
Comment (seanmacavaney) [reply tests/test_experiment_precomp.py:12]: I don’t feel super strongly, but my gut says it’s not sufficiently different than other experiment options that don’t get their own file.

File: pyterrier/pipelines.py [hunk 4]
```diff
@@ -144,6 +145,66 @@ def _ir_measures_to_dict(
         rtr_aggregated[m_name] = metric_agg[m_name].result()
     return rtr_aggregated
 
+def _lcsMany(pipes : List[List[pt.Transformer]]) -> Tuple[List[Transformer], List[List[Transformer]]]:
+    assert len(pipes) > 1
+    
+    lens = [len(p) for p in pipes]
+    common_prefix=[]
+    
+    for i in range(min(lens)):
```
[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier/pipelines.py:154]: I can't imagine the zip version. The lens list could be part of the range(min()), but it made the line rather long

File: pyterrier/pipelines.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier/pipelines.py:156]: Dataframes can't be conflated so we just want to return the no conflation return. Perhaps that can be identified more quickly.

[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier/pipelines.py:156]: Moved dataframe identification to _identifyCommon

File: pyterrier/pipelines.py [hunk 5]
```diff
@@ -333,6 +394,7 @@ def Experiment(
         save_dir : Optional[str] = None,
         save_mode : SAVEMODE_TYPE = 'warn',
         save_format : SAVEFORMAT_TYPE = 'trec',
+        precompute_shared : bool = False,
```
[the remark this answers is not in the document]
Comment (Parry-Parry) [reply pyterrier/pipelines.py:397]: Is equivalence explicitly defined in a transformer, or is it looking for an identical object with default equals behaviour? If this is the case, i.e. no weird backend stuff could break it, then why not make it True?

[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier/pipelines.py:397]: With default equals (i.e. object ref). I just dont want this to complicate (e.g.) IR coursework next semester until we are confident it works properly. Sean and I have discussed committing now for a 0.12.1 release, but not turning on by default until a later release.

File: pyterrier/pipelines.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (cmacdonald) [reply pyterrier/pipelines.py]: Thanks @Parry-Parry - passes my tests too

File: pyterrier/pipelines.py [hunk 6]
```diff
@@ -506,6 +563,48 @@ def _apply_round(measure, value):
...
     }
+    
+    
+    common_pipe, execution_retr_systems = _identifyCommon(retr_systems)
+    precompute_time = 0
+    if precompute_shared and common_pipe is not None: 
+        print("Precomputing results of %d topics on shared pipeline component %s" % (len(topics), str(common_pipe)), file=sys.stderr)
+
+        tqdm_args_precompute = tqdm_args.copy()
+        tqdm_args_precompute['desc'] = "pt.Experiment precomputation"
```
Comment (seanmacavaney) [inline pyterrier/pipelines.py:574]: disable if not verbose?

File: pyterrier/pipelines.py [hunk 7]
```diff
@@ -506,6 +563,48 @@ def _apply_round(measure, value):
...
+            tqdm_args_precompute['total'] = math.ceil((len(topics) / batch_size))
+            with pt.tqdm(**tqdm_args_precompute) as pbar:
+                precompute_results = []
+                for r in common_pipe.transform_gen(topics, batch_size=batch_size):
+                    precompute_results.append(r)
+                    pbar.update(1)
+                execution_topics = pd.concat(precompute_results)
+        else:  
+            tqdm_args_precompute['total'] = 1 
+            with pt.tqdm(**tqdm_args_precompute) as pbar:
```
Comment (seanmacavaney) [inline pyterrier/pipelines.py:591]: Is this tqdm meant to show that the common prefix is running?

File: pyterrier/pipelines.py [hunk 8]
```diff
@@ -506,6 +563,48 @@ def _apply_round(measure, value):
         'total' : len(retr_systems),
         'desc' : 'pt.Experiment'
     }
+    
+    
+    common_pipe, execution_retr_systems = _identifyCommon(retr_systems)
```
Comment (seanmacavaney) [inline pyterrier/pipelines.py:568]: This function is getting a bit unwieldy. Maybe it's worth splitting out this bit into a helper function?

File: pyterrier/pipelines.py (hunk 7 above)
Comment (cmacdonald) [reply pyterrier/pipelines.py:591]: yes

File: pyterrier/pipelines.py (hunk 6 above)
Comment (cmacdonald) [reply pyterrier/pipelines.py:574]: already handled in tqdm_args

Comment (seanmacavaney) [reply pyterrier/pipelines.py:574]: ahhh, ok

**Comment by cmacdonald** [comment]
Final note - the pt.Experiment argument was renamed to `precompute_prefix`

**Pull request merged**